producer/data/ref_counted.go

```diff
@@ -90,19 +90,15 @@ func (d *refCountedData) Drop() bool {
 }
 
 func (d *refCountedData) IsDroppedOrConsumed() bool {
-	d.RLock()
-	r := d.isDroppedOrConsumed
-	d.RUnlock()
-	return r
+	return d.isDroppedOrConsumed.Load()
 }
 
 func (d *refCountedData) finalize(r producer.FinalizeReason) bool {
 	d.Lock()
-	if d.isDroppedOrConsumed {
+	if !d.isDroppedOrConsumed.CAS(false, true) {
```

Comment:
nit: it's probably more readable to write this as
```
if d.isDroppedOrConsumed.Load() {
  d.Unlock()
  return false
}
d.isDroppedOrConsumed.Store(true)
```

Reply:
sure sg